### PR TITLE
[feat] url maintenance

### PR DESCRIPTION
**Pull request by SammyBaek**
## Changes made
This PR maintains the user's path URL when checking for logged-in status.
- Previously, whenever user first load a page with a path (e.g. `/mates`), the app redirects them to login page and then to the dashboard after verifying their auth.
- If they are logged out, it still leads user back to dashboard.

### Before
https://user-images.githubusercontent.com/12789302/139516922-b76f9388-1c31-4422-9d3d-c83ab1429532.mov

### After
https://user-images.githubusercontent.com/12789302/139516930-41ecfd63-1ff0-4b5b-8136-744b736d6347.mov

### After (logged out user)
https://user-images.githubusercontent.com/12789302/139516938-04c568b4-86c5-466f-a0eb-8d0fe862cd3e.mov

### Loader (feature added after previous screen captures)
https://user-images.githubusercontent.com/12789302/139540719-85c15166-edea-4217-95d5-678af42a442a.mov

## Work left to be done
- If a user is logged out, the site routes them to Dashboard after logging in. Maintaining url for logged out user is harder since oAuth makes the user go to `accounts.google.com/...` and then back to our app. Maintaining url for this case would require storing the previous path somewhere locally.

### CODE REVIEW

File: client/.eslintrc.js [hunk 1]
```diff
@@ -40,6 +40,7 @@ module.exports = {
     "import/no-extraneous-dependencies": "off",
     "arrow-body-style": ["error", "as-needed"],
     "no-console": ["warn", { allow: ["warn", "error"] }],
+    "no-debugger": "warn",
```
Comment (SammyBaek) [inline client/.eslintrc.js:43]: typing out `debugger // eslint-disable-line` is very annoying

File: client/package.json [hunk 2]
```diff
@@ -57,6 +57,7 @@
     "@types/react-router-dom": "^5.1.9",
     "@typescript-eslint/eslint-plugin": "^4.31.1",
     "@typescript-eslint/parser": "^4.31.1",
+    "axios-mock-adapter": "^1.20.0",
```
Comment (SammyBaek) [inline client/package.json:60]: Tried mocking `axios` natively and gave up after couple of hours :/

File: client/src/pages/App/App.test.tsx [hunk 3]
```diff
@@ -1,43 +1,95 @@
...
     <Router history={history}>
       <App />
     </Router>
   );
-  const matesElement = screen.getByText("Mates");
-  expect(matesElement).toBeInTheDocument();
+  await waitFor(() => expect(history.location.pathname).toBe("/login"));
+});
+
+test("routes /login if API does not return properly formatted data", async () => {
```
Comment (SammyBaek) [inline client/src/pages/App/App.test.tsx:45]: wrote this mostly to get 100% CC :')

File: client/src/pages/Login/LoginPage.tsx [hunk 4]
```diff
@@ -57,8 +64,12 @@ const LoginPage = ({ loggedIn, setLoggedIn }: LoginPageProps): JSX.Element => {
 
   checkloginCallback();
 
-  // Redirect if authentication is done:
-  if (loggedIn) return <Redirect to="/" />;
+  // Redirect if authentication is done
+  if (loggedIn) {
+    const prevPath = location.state?.prevPath ?? "/";
```
Comment (SammyBaek) [inline client/src/pages/Login/LoginPage.tsx:69]: If user is coming back from logging in through oAuth, `prevPath` is undefined.

Comment (EpicPi) [reply client/src/pages/Login/LoginPage.tsx:69]: https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow#:~:text=it%20is%20requesting.-,state,-Recommended
We can fix that - google oauth provides a way to maintain a bit of state when doing auth.

**Review by EpicPi — CHANGES_REQUESTED** [review]
Nice work! Lets make it maintain url even if google oauth tho.

File: client/src/pages/App/App.test.tsx [hunk 5]
```diff
@@ -1,43 +1,95 @@
-import { render, screen } from "@testing-library/react";
+import { render, screen, waitFor } from "@testing-library/react";
+import axios from "axios";
+import MockAdopter from "axios-mock-adapter";
```
Comment (EpicPi) [inline client/src/pages/App/App.test.tsx:3]: Adopter?

File: client/src/pages/App/App.test.tsx (hunk 3 above)
Comment (EpicPi) [reply client/src/pages/App/App.test.tsx:45]: lol

File: client/src/pages/App/App.test.tsx [hunk 6]
```diff
@@ -1,43 +1,95 @@
...
+test("routes Dashboard if logged in", async () => {
+  const history = createMemoryHistory();
+  render(
+    <Router history={history}>
+      <App />
+    </Router>
+  );
+  await waitFor(() => expect(history.location.pathname).toBe("/"));
+  await waitFor(() =>
+    expect(screen.queryByText("Hi, Paul")).toBeInTheDocument()
```
Comment (EpicPi) [inline client/src/pages/App/App.test.tsx:66]: can we grab string 'paul' from where he's defined?

File: client/src/pages/App/App.tsx [hunk 7]
```diff
@@ -14,25 +14,31 @@ import LoginPage from "../Login/LoginPage";
...
 import NewMatesPage from "../NewMates/NewMatesPage";
 
-const App = (): JSX.Element => {
+/**
+ * Main App page
+ * @returns JSX.Element or null
+ */
+const App = (): JSX.Element | null => {
   const [loggedIn, setLoggedIn] = useState(false);
+  const [checkedLoggedIn, setCheckedLoggedIn] = useState(false);
```
Comment (EpicPi) [inline client/src/pages/App/App.tsx:23]: Would this be better implemented as an enum: loggedIn, notLoggedIn, waiting.

**Review by EpicPi — CHANGES_REQUESTED** [review]
Good work!
Some minor nits.

File: client/src/pages/App/App.test.tsx (hunk 5 above)
Comment (SammyBaek) [reply client/src/pages/App/App.test.tsx:3]: I also can't spell lol, nice catch

File: client/src/pages/App/App.test.tsx (hunk 6 above)
Comment (SammyBaek) [reply client/src/pages/App/App.test.tsx:66]: yep, done

File: client/src/pages/App/App.tsx (hunk 7 above)
Comment (SammyBaek) [reply client/src/pages/App/App.tsx:23]: it's a bit different, but it could work

**Review by EpicPi — APPROVED** [review]
lgtm

**Pull request merged**